#**1.Podcast_Analysis**


##**Introduction**

 2 million podcast reviews for 100k podcasts, updated monthly. This dataset is intended to aid in analysis of text feedback and review data.

## **Objectives for this Part**

Practice working with SQLite datasets.

Practice performing EDA.

Practice applying statistical inference procedures.

Practice visualizing data with Matplotlib & Seaborn.

Practice reading data, performing queries and filtering data using SQL and Pandas.

## **Requirements**

Load the data using SQLite and Pandas.

Perform exploratory data analysis. This should include creating 

statistical summaries and charts, testing for anomalies, checking for correlations and other relations between variables, and other EDA elements.

Perform statistical inference. This should include defining the target population, forming multiple statistical hypotheses and constructing confidence intervals, setting the significance levels, conducting z or t-tests for these hypotheses.

Provide clear explanations in your notebook. Your explanations should inform the reader what you are trying to achieve, what results did you get, and what these results mean.

Provide suggestions about how your analysis can be improved. 



## **Evaluation Criteria**

Adherence to the requirements. How well did you meet the requirements?
Depth of your analysis. Did you just skim the surface or did you explored the dataset in depth?

Visualization quality. Did you use charts effectively to visualize patterns in the data? Are your visualizations properly labeled? Did you use colors effectively? Did you adhere to the principle of proportional ink?

Code quality. Was your code well-structure? Did you use the appropriate levels of abstraction? Did you remove commented out and unused code? Did you adhere to the PEP8?

Code performance. Did you use the suitable algorithms and data structures to solve the problems?

# **2.Libraries and Loading Data**

In [1]:
import pandas as pd
import sqlite3
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
pd.set_option('display.max_columns', 500)

In [77]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from plotly.offline import download_plotlyjs, init_notebook_mode,  iplot
init_notebook_mode(connected=True)

In [2]:
from google.colab import drive
drive.mount("/content/drive")

con = sqlite3.connect("/content/drive/MyDrive/database.sqlite")

Mounted at /content/drive


In [3]:
pd.read_sql("SELECT * FROM sqlite_master WHERE TYPE = 'table' ", con)

type        name    tbl_name  rootpage  \
0  table        runs        runs         2   
1  table    podcasts    podcasts         3   
2  table  categories  categories         5   
3  table     reviews     reviews         7   

                                                 sql  
0  CREATE TABLE runs (\n        run_at text not n...  
1  CREATE TABLE podcasts (\n        podcast_id te...  
2  CREATE TABLE categories (\n        podcast_id ...  
3  CREATE TABLE reviews (\n        podcast_id tex...

In [4]:
df_categories = pd.read_sql_query("SELECT * from categories", con)
df_podcasts= pd.read_sql_query("SELECT * from podcasts", con)
df_reviews = pd.read_sql_query("SELECT * from reviews", con)
df_runs = pd.read_sql_query("SELECT * from runs", con)

**Checking the Tables:**

In [5]:
df_categories.head()

podcast_id              category
0  c61aa81c9b929a66f0c1db6cbe5d8548                  arts
1  c61aa81c9b929a66f0c1db6cbe5d8548  arts-performing-arts
2  c61aa81c9b929a66f0c1db6cbe5d8548                 music
3  ad4f2bf69c72b8db75978423c25f379e                  arts
4  ad4f2bf69c72b8db75978423c25f379e           arts-design

In [6]:
df_podcasts.head()

podcast_id   itunes_id  \
0  a00018b54eb342567c94dacfb2a3e504  1313466221   
1  a00043d34e734b09246d17dc5d56f63c   158973461   
2  a0004b1ef445af9dc84dad1e7821b1e3   139076942   
3  a00071f9aaae9ac725c3a586701abf4d  1332508972   
4  a000a500f06555f81220c3eb641aded7  1544900779   

                                    slug  \
0                         scaling-global   
1  cornerstone-baptist-church-of-orlando   
2            mystery-dancing-in-the-dark   
3                      kts-money-matters   
4       word-on-the-street-w-dreak-swift   

                                          itunes_url  \
0  https://podcasts.apple.com/us/podcast/scaling-...   
1  https://podcasts.apple.com/us/podcast/cornerst...   
2  https://podcasts.apple.com/us/podcast/mystery-...   
3  https://podcasts.apple.com/us/podcast/kts-mone...   
4  https://podcasts.apple.com/us/podcast/word-on-...   

                                   title  
0                         Scaling Global  
1  Cornerstone Baptist Church of Orlando  
2           Mystery: Dancing in the Dark  
3                      KTs Money Matters  
4      Word on the Street w/ Dreak Swift

In [7]:
df_reviews.head()

podcast_id  \
0  c61aa81c9b929a66f0c1db6cbe5d8548   
1  c61aa81c9b929a66f0c1db6cbe5d8548   
2  ad4f2bf69c72b8db75978423c25f379e   
3  ad4f2bf69c72b8db75978423c25f379e   
4  ad4f2bf69c72b8db75978423c25f379e   

                                              title  \
0                               really interesting!   
1  Must listen for anyone interested in the arts!!!   
2                                 nauseatingly left   
3                                   Diverse stories   
4                                              👍👍👍👍   

                                             content  rating        author_id  \
0  Thanks for providing these insights.  Really e...       5  F7E5A318989779D   
1  Super excited to see this podcast grow. So man...       5  F6BF5472689BD12   
2  I'm a liberal myself, but its pretty obvious a...       1  1AB95B8E6E1309E   
3  I find Tedx talks very inspirational but I oft...       5  11BB760AA5DEBD1   
4                I love this podcast, it is so good.       5  D86032C8E57D15A   

                  created_at  
0  2018-04-24T12:05:16-07:00  
1  2018-05-09T18:14:32-07:00  
2  2019-06-11T14:53:39-07:00  
3  2018-05-31T13:08:09-07:00  
4  2019-06-19T13:56:05-07:00

In [8]:
df_runs.head()

run_at  max_rowid  reviews_added
0  2021-05-10 02:53:00    3266481        1215223
1  2021-06-06 21:34:36    3300773          13139
2  2021-07-02 18:04:55    3329699          11561
3  2021-08-01 17:54:42    3360315          11855
4  2021-09-02 18:00:30    3390165          11714

**Basic Checks**

In [9]:
df_categories.describe()

podcast_id         category
count                             211760           211760
unique                            109602              110
top     be905117ff9b3a13b1a065801241903b  society-culture
freq                                   8            18698

Table categories consists of two features, representing podcast ID and the category. 

Trying to find out any missing and duplicate values

In [10]:
df_categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211760 entries, 0 to 211759
Data columns (total 2 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   podcast_id  211760 non-null  object
 1   category    211760 non-null  object
dtypes: object(2)
memory usage: 3.2+ MB


Findings:

There are no missing values

In [11]:
df_categories.duplicated().sum()

0

Findings:

There are no duplicated  values

I am trying to see what are the category types

In [12]:
pd.set_option('display.max_rows', 150)
pd.read_sql_query('''
                   SELECT DISTINCT category 
                   FROM categories
                   ''', con)
        
        
        

category
0                                 arts
1                 arts-performing-arts
2                                music
3                          arts-design
4                            education
5                      society-culture
6                     arts-visual-arts
7                              leisure
8                      leisure-hobbies
9                           arts-books
10                          technology
11   society-culture-personal-journals
12       society-culture-relationships
13                             fiction
14             fiction-science-fiction
15                           arts-food
16         education-language-learning
17                       fiction-drama
18                          true-crime
19                      health-fitness
20            health-fitness-nutrition
21                             history
22                              comedy
23                 arts-fashion-beauty
24                             tv-film
25       society-culture-places-travel
26            comedy-comedy-interviews
27                         kids-family
28        health-fitness-mental-health
29               religion-spirituality
30                            business
31                                news
32                  news-business-news
33                   education-courses
34          society-culture-philosophy
35                 leisure-video-games
36                tv-film-film-reviews
37                       leisure-games
38                        spirituality
39                tv-film-film-history
40                    business-careers
41           business-entrepreneurship
42                             science
43              science-earth-sciences
44                       comedy-improv
45                  tv-film-tv-reviews
46            health-fitness-sexuality
47                       news-politics
48                 leisure-home-garden
49             leisure-animation-manga
50              fiction-comedy-fiction
51                        christianity
52                          government
53                              sports
54                            hinduism
55              music-music-commentary
56                  business-marketing
57              music-music-interviews
58                    education-how-to
59             science-social-sciences
60         society-culture-documentary
61                 business-management
62                      news-tech-news
63          education-self-improvement
64                   sports-wilderness
65   health-fitness-alternative-health
66             health-fitness-medicine
67                      leisure-crafts
68                  business-investing
69            science-natural-sciences
70                      science-nature
71                 tv-film-after-shows
72                       sports-soccer
73            kids-family-pets-animals
74                             judaism
75                            religion
76                     sports-football
77               kids-family-parenting
78              health-fitness-fitness
79             news-entertainment-news
80                     news-daily-news
81        kids-family-stories-for-kids
82             tv-film-film-interviews
83                 music-music-history
84               science-life-sciences
85                     comedy-stand-up
86                    leisure-aviation
87                  leisure-automotive
88                         sports-golf
89                 business-non-profit
90                               islam
91                news-news-commentary
92                    news-sports-news
93                   sports-basketball
94                     sports-baseball
95               sports-fantasy-sports
96      kids-family-education-for-kids
97                   science-astronomy
98                     science-physics
99                   science-chemistry
100                     sports-running
101                   sports-wrestling
102          

There are 110 unique categories in a table.However most of them are similiar. 
Therefore below categories are narrowed down by combining similiar ones ito one.



In [13]:
#write alterntive
query = """
select distinct(category_updated)
from(
  select *, 
   (case
      when category like "%art%" then 'Arts'
      when category like "%business%" then 'Business'
      when category like "%culture%" then 'Culture'
      when category = "buddhism" then 'Religion'
      when category = "hinduism" then 'Religion'
      when category = "islam" then 'Religion'
      when category = "christianity" then 'Religion'
      when category = "judaism" then 'Religion'
      when category = "religion-spirituality" then 'Religion'
      when category like "%society%" then "Society"
      when category like "%Fict%" then "Fiction"
      when category like "%Edu%" then "Education"
      when category like "%Health%" then "Health"
      when category like "%Tv%" then "Tv"
      when category like "%Comedy%" then "Comedy"
      when category like "%Kids%" then "Kids"
      when category like "%Leisure%" then "Leisure"
      when category like "%News%" then "News"
      when category like "%Music%" then "Music"
      when category like "%Science%" then "Science"
      when category like "%sport%" then "sports"
        else upper(substr(category, 1, 1)) || substr(category, 2)
   end) AS category_updated
  from categories
)
"""
df=pd.read_sql(query, con)
df

category_updated
0              Arts
1             Music
2         Education
3           Culture
4           Leisure
5        Technology
6           Fiction
7        True-crime
8            Health
9           History
10           Comedy
11               Tv
12             Kids
13         Religion
14         Business
15             News
16     Spirituality
17          Science
18       Government
19           sports

Findings:

After the extraction there exist only 20 different categories, each representing a different field.

In [14]:
df

category_updated
0              Arts
1             Music
2         Education
3           Culture
4           Leisure
5        Technology
6           Fiction
7        True-crime
8            Health
9           History
10           Comedy
11               Tv
12             Kids
13         Religion
14         Business
15             News
16     Spirituality
17          Science
18       Government
19           sports

**podcasts basic checks**

In [15]:
df_podcasts.sample(5)

podcast_id   itunes_id  \
16595  ae76297fc07e8ea93eed09b5c5cc71f7  1510280092   
78711  e4e247bca2165b58dbee1f00b0e98f8b   968669365   
29983  ba20ba98c4e95bfc947bcc276b1ca537  1310581368   
76527  e2f637cbe0c462cb004770a568eefe2b  1569227177   
52693  ce02153c484743b36e023d79bc835285  1539059917   

                                                 slug  \
16595                                     black-angel   
78711  sigtalks-sigafoose-the-chiropractic-philosophy   
29983                       restaurant-report-podcast   
76527                                       soul-deep   
52693                     ti-ch%C3%A8z-ba-the-podcast   

                                              itunes_url  \
16595  https://podcasts.apple.com/us/podcast/black-an...   
78711  https://podcasts.apple.com/us/podcast/sigtalks...   
29983  https://podcasts.apple.com/us/podcast/restaura...   
76527  https://podcasts.apple.com/us/podcast/soul-dee...   
52693  https://podcasts.apple.com/us/podcast/ti-ch%C3...   

                                                 title  
16595                                     Black Angel🖤  
78711  SigTalks: Sigafoose THE chiropractic philosophy  
29983                        Restaurant Report Podcast  
76527                                        Soul Deep  
52693                           Ti Chèz Ba The Podcast

In [16]:
df_podcasts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 109602 entries, 0 to 109601
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   podcast_id  109602 non-null  object
 1   itunes_id   109602 non-null  int64 
 2   slug        109602 non-null  object
 3   itunes_url  109602 non-null  object
 4   title       109602 non-null  object
dtypes: int64(1), object(4)
memory usage: 4.2+ MB


Podcasts does not have any missing values.

In [17]:
df_podcasts[df_podcasts.title.astype(str).str.contains('None')]


podcast_id   itunes_id  \
13670  ac086204f34c0b65f93c460027cd1fad  1558021437   
15073  ad3316bd30d07b5878f2f61b3b951ff2  1337885213   
16113  ae140141b4ae629223ff4aa0497a7f82  1347751328   
31089  bb2396c22f4b4b919321b6f40d0d9427  1292426978   
33140  bd0117a0c07adde7f444b8e2eb073955  1451821858   
38469  c19de4a016b98d5fe91af3f863eba319   903699599   
43479  c60c77815e6782f22e9f49da11584efd   593751261   
48307  ca3d2d0cf4587058a25ef043664340fc  1514036779   
55519  d07ae5e7ab84921310eb560790346065  1439588586   
58165  d2cc76a7476b1e0b22255ba3408ee95b  1433153093   
63161  d73338bbb9fc1e00bd020082a8b3401a  1551752797   
65900  d9a80cfa78efbc37ce95cb4b47646b66  1114394782   
75242  e1d77855e217eb37e0ab95b51f9f6bf8  1557086266   
83155  e8ccfeba08bc471f5e963f20d9f2a52a  1558064647   
94016  f2463d15fe2aa1792aef325aed045b0f  1457092789   

                                                    slug  \
13670                               none-of-this-is-reel   
15073                             dungeon-master-of-none   
16113                                       nonexclusive   
31089                                         none-sense   
33140                              none-of-your-business   
38469  thus-spake-zarathustra-book-for-all-none-by-fr...   
43479                            the-none-to-run-podcast   
48307                      none-of-your-business-podcast   
55519                               none-of-this-is-real   
58165                                       none-of-that   
63161                       none-of-this-makes-any-sense   
65900                                           bar-none   
75242                                     second-to-none   
83155                                 malice-toward-none   
94016                                       hole-in-none   

                                              itunes_url  \
13670  https://podcasts.apple.com/us/podcast/none-of-...   
15073  https://podcasts.apple.com/us/podcast/dungeon-...   
16113  https://podcasts.apple.com/us/podcast/nonexclu...   
31089  https://podcasts.apple.com/us/podcast/none-sen...   
33140  https://podcasts.apple.com/us/podcast/none-of-...   
38469  https://podcasts.apple.com/us/podcast/thus-spa...   
43479  https://podcasts.apple.com/us/podcast/the-none...   
48307  https://podcasts.apple.com/us/podcast/none-of-...   
55519  https://podcasts.apple.com/us/podcast/none-of-...   
58165  https://podcasts.apple.com/us/podcast/none-of-...   
63161  https://podcasts.apple.com/us/podcast/none-of-...   
65900  https://podcasts.apple.com/us/podcast/bar-none...   
75242  https://podcasts.apple.com/us/podcast/second-t...   
83155  https://podcasts.apple.com/us/podcast/malice-t...   
94016  https://podcasts.apple.com/us/podcast/hole-in-...   

                                                   title  
13670                               None of This is Reel  
15073                             Dungeon Master of None  
16113                                       Nonexclusive  
31089                                         None Sense  
33140                              None Of Your Business  
38469  Thus Spake Zarathustra: A Book for All and Non...  
43479                            The None to Run Podcast  
48307                      None Of Your Business Podcast  
55519                               None of This is Real  
58165                                       None of That  
63161                       None of This Makes Any Sense  
65900                                           Bar None  
75242                                     SECond To None  
83155                                 Malice Toward None  
94016                                      Hole  In None

In [18]:
df_podcasts.describe()

itunes_id
count  1.096020e+05
mean   1.308588e+09
std    3.576728e+08
min    7.332927e+07
25%    1.225780e+09
50%    1.467143e+09
75%    1.528596e+09
max    1.658655e+09

checking any duplicate values in the podcasts data

In [19]:
df_podcasts.duplicated().any()

False

In [20]:

def count_unique_values(df):
    return df.nunique()

df = pd.DataFrame(df_podcasts)
print(count_unique_values(df))


podcast_id    109602
itunes_id     109602
slug          108503
itunes_url    109602
title         108857
dtype: int64


It appears that there may be duplicates present in the 'slug' and 'title' columns of the DataFrame. To further investigate, I will examine if these duplicates indicate that certain podcasts have been included multiple times in the dataset or if they simply indicate that certain podcasts have similar 'slug' and 'title' values but different overall content.

In [21]:
duplicate_slugs = df_podcasts[df.duplicated(subset='slug', keep=False)]['slug']
duplicate_slugs.value_counts()


lets-talk-about-it       9
no-title                 8
in-the-loop              6
whats-next               6
off-the-cuff             6
                        ..
what-could-go-right      2
real-talk-with-raquel    2
song-story               2
shower-thoughts          2
to-the-point             2
Name: slug, Length: 945, dtype: int64

In [22]:
duplicate_slugs.value_counts()['off-the-record']


4

In [23]:
multiple_slugs = df_podcasts.query("slug == 'in-the-loop'")[['slug', 'itunes_url','title']]
multiple_slugs


slug                                         itunes_url  \
36557   in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   
51645   in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   
52350   in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   
61712   in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   
74918   in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   
100681  in-the-loop  https://podcasts.apple.com/us/podcast/in-the-l...   

              title  
36557   In The Loop  
51645   In the Loop  
52350   In The Loop  
61712   In the Loop  
74918   In the Loop  
100681  In The Loop

Findings:

the Two podcasts have the same 'slug' or 'title', they may refer to different instances of podcasts with the same name. For example, they could be produced by different authors or cover different subjects. It's also possible that they are different episodes of the same podcast series. we also further investigated the duplicate entries and it is false


**Reviews basic checks**

In [24]:
df_reviews.sample(5)

podcast_id  \
162762   d49724209c282a0edf4b18d08b2df29e   
1597757  cff81f05e5c948cc7fa2f494d77010a6   
197848   cc7b7b176f20cb9735ce58cf4cee751c   
1067191  f529487b091b7659049b17420ae20267   
337839   c6e025d501ac8cc015870e2ddefd59f5   

                                              title  \
162762                              Awesome Podcast   
1597757  Perfect mix of entertaining and insightful   
197848              Old school is the real D&D Next   
1067191     Enlightening and spiritually awakening!   
337839                                  Good listen   

                                                   content  rating  \
162762   Ernesta is committed to bringing people great ...       5   
1597757  Definitely my new favorite podcast. It is so e...       5   
197848   I've been listening to RFI for a few years and...       5   
1067191  I absolutely love this podcast. I feel so much...       5   
337839   If you are tired of hearing “r”s pronounced at...       5   

               author_id                 created_at  
162762   1519834B69FB8FD  2019-05-15T09:39:31-07:00  
1597757  54061B137BA1562  2020-12-14T12:33:06-07:00  
197848   D4A00D9430977E7  2012-03-24T14:11:46-07:00  
1067191  F587B7BB7894DED  2020-09-07T10:27:56-07:00  
337839   247BD6A6ECFBCBF  2018-05-17T16:10:57-07:00

In [25]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2025333 entries, 0 to 2025332
Data columns (total 6 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   podcast_id  object
 1   title       object
 2   content     object
 3   rating      int64 
 4   author_id   object
 5   created_at  object
dtypes: int64(1), object(5)
memory usage: 92.7+ MB


We can see that all of the columns exept of 'rating' is categorial. Rating is of numeric type.

Lets find out any duplicates found

In [26]:
df_reviews.duplicated().sum()

652

We can see that there are 652 duplicated entries

In [27]:
df_reviews[df_reviews.duplicated(keep=False)]

podcast_id  \
3269     d31472739beeec4c8593810f7c158ad5   
3275     d31472739beeec4c8593810f7c158ad5   
6083     ad237e58eec063d15c0856131131c4e2   
6114     ad237e58eec063d15c0856131131c4e2   
13835    e8fb77c98183e85b6bf75e85da02892d   
...                                   ...   
1946650  ae97c70120af155f0813f5fba880e60a   
1956830  a0dcc13fd50bb25c866a8b9a2eca0f50   
1964345  f2970432aad718a000ebac10e48ae6b0   
1970374  b41b7a2e79c1f747d21babca3de1c184   
1979442  f416dbd6e2740cacc5bec1a37de8f3d4   

                                                  title  \
3269                                    Great storyline   
3275                                    Great storyline   
6083     Horrifying immersive well written storytelling   
6114     Horrifying immersive well written storytelling   
13835    If you like Myths and Legends, look no further   
...                                                 ...   
1946650                           “Chyna,Chyna,Chyna..”   
1956830                                          Racist   
1964345                                  Awe Inspiring!   
1970374                                Awesome Podcast!   
1979442                   History taught on a new level   

                                                   content  rating  \
3269     Love this exciting story, delving into mysteri...       5   
3275     Love this exciting story, delving into mysteri...       5   
6083     I love S&S and Dark Topic.  West side Fairytai...       5   
6114     I love S&S and Dark Topic.  West side Fairytai...       5   
13835    Myths and Legends’ Jason Weiser brings his mod...       5   
...                                                    ...     ...   
1946650  The Biden administration is bought and paid fo...       5   
1956830  Continues to mispronounce foreign golfer’s nam...       1   
1964345  Between this podcast and Spooked my mind and l...       5   
1970374  New listener! Really enjoy the podcast, the Tu...       5   
1979442  Love this podcast, when people ask me what my ...       5   

               author_id                 created_at  
3269     534676FEA811BBA  2017-07-13T22:05:51-07:00  
3275     534676FEA811BBA  2017-07-13T22:05:51-07:00  
6083     F19703FACB7CA26  2019-02-22T13:06:34-07:00  
6114     F19703FACB7CA26  2019-02-22T13:06:34-07:00  
13835    213BE0AEEC056FB  2018-09-09T21:53:40-07:00  
...                  ...                        ...  
1946650  1079EE1C386A625  2022-08-07T16:11:20-07:00  
1956830  3A41905452F889F  2022-07-26T03:52:22-07:00  
1964345  CBB1165EE6429D9  2022-05-27T10:50:56-07:00  
1970374  0837294EC18BD32  2022-07-12T14:49:18-07:00  
1979442  F3B1CD4684D7177  2022-05-28T20:04:45-07:00  

[1231 rows x 6 columns]

In [28]:
#merging
# merged_df = df_reviews.merge(df_podcasts, on='podcast_id', how='left')
# merged_df[merged_df['podcast_id'].isna()]

x=pd.read_sql_query(
    ''' 
      SELECT DISTINCT r.podcast_id,
             p.title AS Title
      FROM reviews r 
      LEFT JOIN podcasts p
      ON p.podcast_id = r.podcast_iD
      WHERE p.podcast_id IS NULL
    ''', con
)
x.count()

podcast_id    1523
Title            0
dtype: int64

Findings:

It appears that there are 1523 podcasts that are not present in the "podcasts" table. This could be due to the fact that the table was not updated with the relevant information for these specific podcasts during the creation of the database. This can happen when there is a mismatch between the data being added to the table and the source of data.

**Runs basic checks**

In [29]:
df_runs.sample(5)

run_at  max_rowid  reviews_added
8   2022-06-22 19:44:32    4479220          29146
12  2022-10-01 21:56:38    5364248          16649
13  2022-11-03 21:12:30    5427586          22495
0   2021-05-10 02:53:00    3266481        1215223
14  2022-12-10 01:17:07    5488314          21333

In [30]:
df_runs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   run_at         15 non-null     object
 1   max_rowid      15 non-null     int64 
 2   reviews_added  15 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 488.0+ bytes


We can see that run_at is a object but it should be time series and other features were numeric

In [31]:
df_runs.reviews_added.sum()

2025333

The table shows a record of updates made to the "reviews_df" table. The "reviews_added" column represents the number of reviews that were added to the table during each update

## 3.Exploratory Data Analysis

**Analysis of the categories**

checking the distinct categories

In [78]:
import plotly.express as px
df = pd.read_sql('SELECT * FROM categories', con)
df['category_updated'] = df['category'].apply(lambda x: 'Arts' if 'art' in x.lower() else
                                                           'Business' if 'business' in x.lower() else
                                                           'Culture' if 'culture' in x.lower() else
                                                           'Religion' if x in ['buddhism', 'hinduism', 'islam', 'christianity', 'judaism', 'religion-spirituality'] else
                                                           'Society' if 'society' in x.lower() else
                                                           'Fiction' if 'fict' in x.lower() else
                                                           'Education' if 'edu' in x.lower() else
                                                           'Health' if 'health' in x.lower() else
                                                           'Tv' if 'tv' in x.lower() else
                                                           'Comedy' if 'comedy' in x.lower() else
                                                           'Kids' if 'kids' in x.lower() else
                                                           'Leisure' if 'leisure' in x.lower() else
                                                           'News' if 'news' in x.lower() else
                                                           'Music' if 'music' in x.lower() else
                                                           'Science' if 'science' in x.lower() else
                                                           'Sports' if 'sport' in x.lower() else
                                                           x.capitalize() if x.isalpha() else x)

result = df.groupby(['category_updated']).agg({'category_updated': 'count'})
result.rename(columns={'category_updated': 'POPULARITY'}, inplace=True)
result = result.reset_index()
result = result.sort_values(by='POPULARITY', ascending=False)
result = result.reset_index(drop=True)

fig = px.bar(result, x='category_updated', y='POPULARITY')
iplot(fig)



In [66]:
)pip install -U kaleido

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 13.5 MB/s eta 0:00:00


Findings:

The most common category of the podcast is **'culture'** and the less common category is **'government'** 

We can now check the average ratings of podcasts in each category.


In [33]:
query = """
        select c.category_updated, round(avg(rating),2) ratings, count(*) num_reviews
        from reviews 
        join (select *,
              (case
                  when category like "%art%" then 'Arts'
                  when category like "%business%" then 'Business'
                  when category like "%culture%" then 'Culture'
                  when category = "buddhism" then 'Religion'
                  when category = "hinduism" then 'Religion'
                  when category = "islam" then 'Religion'
                  when category = "christianity" then 'Religion'
                  when category = "judaism" then 'Religion'
                  when category = "religion-spirituality" then 'Religion'
                  when category like "%society%" then "Society"
                  when category like "%Fict%" then "Fiction"
                  when category like "%Edu%" then "Education"
                  when category like "%Health%" then "Health"
                  when category like "%Tv%" then "Tv"
                  when category like "%Comedy%" then "Comedy"
                  when category like "%Kids%" then "Kids"
                  when category like "%Leisure%" then "Leisure"
                  when category like "%News%" then "News"
                  when category like "%Music%" then "Music"
                  when category like "%Science%" then "Science"
                  when category like "%sport%" then "sports"
                  else upper(substr(category, 1, 1)) || substr(category, 2)
              end) as category_updated
            FROM categories) c
        using (podcast_id)
        group by category_updated
        order by rating desc
        """
avg_ratings = pd.read_sql(query, con)
avg_ratings = avg_ratings.sort_values(by='ratings', ascending=False)
avg_ratings= avg_ratings.reset_index(drop=True)
avg_ratings


category_updated  ratings  num_reviews
0      Spirituality     4.86        45499
1          Business     4.85       433409
2          Religion     4.83       238226
3         Education     4.82       364181
4            Health     4.79       330365
5             Music     4.78        88901
6           Leisure     4.76       188811
7              Kids     4.75       154576
8              Arts     4.73       252219
9            sports     4.64       310198
10           Comedy     4.64       405100
11          Fiction     4.61        61506
12          Culture     4.58       646583
13               Tv     4.54       247168
14       Technology     4.53        47286
15          History     4.53        58100
16          Science     4.52        73552
17       Government     4.43        15286
18             News     4.33       316113
19       True-crime     4.17       157999

In [34]:
fig = px.bar(avg_ratings, x='category_updated', y='ratings', color='num_reviews',log_y=True,
             title='Average Ratings by Category',
             labels={'category_updated': 'Category', 'average_rating': 'Average Rating', 'num_reviews': 'Number of Reviews'})
fig.show()


Findings:

We can see that podcasts of categories Spirituality and Businesss are rated the highest, while podcasts about News and Crime are rated the lowest.

Now we can see most popular podcasts

In [35]:
query = """
        select c.category_updated,  count(*) as count
        from podcasts
        join (select *,
              (case
                  when category like "%art%" then 'Arts'
                  when category like "%business%" then 'Business'
                  when category like "%culture%" then 'Culture'
                  when category = "buddhism" then 'Religion'
                  when category = "hinduism" then 'Religion'
                  when category = "islam" then 'Religion'
                  when category = "christianity" then 'Religion'
                  when category = "judaism" then 'Religion'
                  when category = "religion-spirituality" then 'Religion'
                  when category like "%society%" then "Society"
                  when category like "%Fict%" then "Fiction"
                  when category like "%Edu%" then "Education"
                  when category like "%Health%" then "Health"
                  when category like "%Tv%" then "Tv"
                  when category like "%Comedy%" then "Comedy"
                  when category like "%Kids%" then "Kids"
                  when category like "%Leisure%" then "Leisure"
                  when category like "%News%" then "News"
                  when category like "%Music%" then "Music"
                  when category like "%Science%" then "Science"
                  when category like "%sport%" then "sports"
                  else upper(substr(category, 1, 1)) || substr(category, 2)
              end) as category_updated
            FROM categories) c
        using (podcast_id)
        group by 1
        order by 2 desc
        """
avg_ratings = pd.read_sql(query, con)

avg_ratings

category_updated  count
0           Culture  27536
1          Business  22855
2          Religion  21112
3         Education  20938
4              Arts  16185
5            Comedy  14548
6            Health  14504
7            sports  13138
8           Leisure  12023
9                Tv  10794
10             News  10290
11            Music   7807
12             Kids   4605
13          Science   3289
14       Technology   3227
15          Fiction   2762
16     Spirituality   2297
17          History   1657
18       True-crime   1255
19       Government    938

In [36]:
fig = px.bar(avg_ratings, x='category_updated', y='count',title='Popular Podcasts',)
fig.show()

Findings:

The most popular categories of the podcasts are Culture and Business

**Analysis of the Podcasts**

We can see now most populat tiltle with number of reviews and ratings

In [37]:
query = """select p.title, count(*) num_reviews, avg(rating) average_rating
          from podcasts p
          join reviews r
          using (podcast_id)
          group by 1
          order by 2 desc
          limit 10
        """
most_reviews = pd.read_sql(query, con)
most_reviews

title  num_reviews  \
0                                       Crime Junkie        32474   
1  My Favorite Murder with Karen Kilgariff and Ge...        10551   
2                                   Wow in the World         9470   
3                               The Ben Shapiro Show         8130   
4                                True Crime Obsessed         7214   
5                                      Story Pirates         7186   
6                                          The Daily         5367   
7                Know Your Aura with Mystic Michaela         5011   
8                                    Crime in Sports         4974   
9             FantasyPros - Fantasy Football Podcast         4615   

   average_rating  
0        4.283981  
1        3.611790  
2        4.777719  
3        3.809840  
4        4.215969  
5        4.735597  
6        3.350289  
7        4.983436  
8        4.929232  
9        4.821235

In [38]:

fig = px.bar(most_reviews, x='title', y='num_reviews', color='average_rating',log_y=True,
             title='Podcasts with the Most Reviews',
             labels={'title': 'Podcast Title', 'num_reviews': 'Number of Reviews',
                     'average_rating': 'Average Rating'})
fig.show()


Findings:

It seems that "My Favourite Murder with Karen Kilgariff and Crime Junkie" is the most reviewd podcsat having more than twice reviews compared to other podcasts. 

We can also see the podcasts with highest number of rating with less counts 

In [39]:
query = """select p.title, round(avg(rating), 2) rating, count(*) num_reviews
          from reviews r
          join podcasts p
          using (podcast_id)
          group by 1
          having num_reviews > 1000
          order by 2 desc
          limit 10
        """
top_podcasts= pd.read_sql(query, con)
top_podcasts

title  rating  num_reviews
0                             Shawn Ryan Show    4.99         1800
1                                Sassy Lashes    4.99         1150
2         Know Your Aura with Mystic Michaela    4.98         5011
3                     Daebak Show w/ Eric Nam    4.98         2981
4                   Biceps After Babies Radio    4.98         1280
5                              BASS TALK LIVE    4.96         1562
6  Andrea Savage: A Grown-Up Woman #buttholes    4.96         1325
7                          Healthier Together    4.95         1336
8                                Ghostrunners    4.95         1060
9                             Fantastic Geeks    4.95         1624

In [40]:
fig = px.bar(top_podcasts.sort_values(by='rating', ascending=False),
             x='title', y='rating', color='num_reviews',
             log_y=True,
             title='Top Podcasts by Average Rating and Number of Reviews',
             labels={'title': 'Podcast Title', 'num_reviews': 'Number of Reviews',
                     'rating': 'Average Rating'})
fig.show()

"Shawn Ryan Show" and 'Sassy Lashes' seems to be the best rated podcast 

In [41]:
best_podcasts=pd.read_sql_query('''
        SELECT p.title AS Title, AVG(r.rating) AS Average_rating, COUNT(r.rating) AS Number_of_votes
        FROM podcasts p
        JOIN reviews r
        ON p.podcast_id = r.podcast_id
        GROUP BY p.podcast_id
        ORDER BY 2 DESC
 



              ''', con)
best_podcasts.head(1)
best_podcasts.tail(1)

Title  Average_rating  Number_of_votes
109601  Mystery: Dancing in the Dark             1.0                1

Now try to find out best and worst podcasts

In [42]:
best_podcasts=pd.read_sql_query('''
        SELECT p.title AS Title, AVG(r.rating) AS Average_rating, COUNT(r.rating) AS Number_of_votes
        FROM podcasts p
        JOIN reviews r
        ON p.podcast_id = r.podcast_id
        GROUP BY p.podcast_id
        HAVING COUNT(r.rating) >= 500
        ORDER BY 2 DESC
        LIMIT 10



              ''', con)
best_podcasts

Title  Average_rating  \
0                    I Think You're Dope w/ Eric Nam        5.000000   
1  THE STEFANIE GASS SHOW - Christian Entrepreneu...        4.998028   
2        Above Average Podcast with Travis and Jesse        4.996516   
3  Discover Your Talent–Do What You Love | Build ...        4.996071   
4                                       Sassy Lashes        4.993043   
5                        Making Podcasts Great Again        4.992844   
6            FILL IN THE BLANK with Carlos Whittaker        4.991653   
7                          The Hustle Sanely Podcast        4.989831   
8                                    Shawn Ryan Show        4.989444   
9                                        The Lo Life        4.985353   

   Number_of_votes  
0              538  
1              507  
2              574  
3              509  
4             1150  
5              559  
6              599  
7              590  
8             1800  
9              751

Now checking for the worst podcasts

In [43]:
worst_podcasts=pd.read_sql_query('''
                  SELECT p.title AS Title, AVG(r.rating) AS Average_rating, COUNT(r.rating) AS Number_of_votes
                  FROM podcasts p
                  JOIN reviews r
                  ON p.podcast_id = r.podcast_id
                  GROUP BY p.podcast_id
                  HAVING COUNT(r.rating) >= 500
                  ORDER BY AVG(r.rating) ASC
                  LIMIT 10

                 ''', con)
worst_podcasts


Title  Average_rating  Number_of_votes
0                          Murder in Illinois        2.380952              525
1                               It's Me, Tinx        2.554348              552
2                                     Missing        2.686728              648
3  The Ben and Ashley I Almost Famous Podcast        2.765871             1969
4                       The Fighter & The Kid        2.768764             1812
5             Click Bait with Bachelor Nation        2.770445              697
6                             Atlanta Monster        2.895770              662
7                   This Week in Tech (Audio)        2.944816              598
8                             Cold Case Files        2.952416             1345
9                              Betches Brides        2.976266              632

In [44]:
fig = px.bar(best_podcasts, x='Title', y='Average_rating', color='Average_rating', log_y=True,title='Best Podcasts')
fig2 = px.bar(worst_podcasts, x='Title', y='Average_rating', color='Average_rating', log_y=True, title='Worst Podcasts')

fig.show()
fig2.show()

Findings:


To have a good understanding of what podcast are mostly liked/disliked, I had to not only look at the average score of the reviews but also at the total number of voters. 

**Analysis of the Reviews**

Finding out about any reviewers voted more than 100 times?


In [45]:
pd.read_sql_query( '''
       
                SELECT DISTINCT author_id As Author_ID,
                      COUNT(author_id) AS Number_of_reviews
                FROM reviews
                GROUP BY 1
                HAVING COUNT(author_id) >= 100
                ORDER BY 2 DESC
                ''', con)

Author_ID  Number_of_reviews
0   D3307ADEFFA285C                615
1   791F376ACA3C08D                615
2   96A3F0358E7808D                475
3   04C3B6DEFE22FB1                431
4   8858ED4249110B6                397
5   7F9EA8B1A449BFC                361
6   4B25015E4D7A1DA                341
7   B20DB44F45DB97A                323
8   A52EBC918BC19D1                321
9   184F09732A68A08                310
10  C02C514DAA9D885                309
11  6033FF110F6CD23                302
12  8EA31F030185A88                293
13  3D53C30E1911630                288
14  8843FBA0937364E                254
15  8167974065A0260                246
16  3E2C917B8E60AB3                192
17  04E85BADC6E8986                184
18  D42837C5F8BD234                177
19  C0EA806D516803B                177
20  675A82764372546                168
21  61ED6D289B34165                164
22  E1E3D5BF8BDF567                163
23  B67D7BE0E454268                150
24  871BCE7CB360EBB                148
25  2975ECC1C888139                144
26  01B41DDC59E7F7F                137
27  59223EEA42FBFCE                133
28  B3BCE599817B315                132
29  6DBBFD322D38BD2                127
30  CBF46E5BEDF2F9A                125
31  C6ACA659087DCC1                125
32  FB2467731CA599C                124
33  426DDBF8D527E3A                124
34  E97FCC8DA48F89E                123
35  DCD872F1EE93AD5                121
36  B53119FB3E276D3                119
37  961F6822AAA4F4D                116
38  2899EB23FFF842C                116
39  304D4EE7DF74535                115
40  94F11F85A9AD087                111
41  B1877C1A2DEDAB4                110
42  44C5207F0E59BA5                109
43  CBFA7DE2DA5D1D1                108
44  241D91E70A625DB                108
45  C9EDC715E2AEA2F                107
46  E32005224E7097C                106
47  AC072D3BF1CB749                106
48  1AED5583C51E014                105
49  29A468BE322892B                103
50  DE6E19493E35CE1                101
51  D1477CF898A1779                101

Findings:

There are 51 bvery active reviewers on iTunes platform.

I would like to find out if specific author comented on a specific podcast more than once.

In [46]:
pd.read_sql_query(
        '''
SELECT author_id, COUNT(author_id) AS Number_of_reviews, p.title AS Title, ROUND(AVG(rating), 2) AS Average_rating
FROM reviews r
JOIN podcasts p 
ON r.podcast_id = p.podcast_id
JOIN (SELECT podcast_id, GROUP_CONCAT(category, ', ') AS All_categories
      FROM categories
      GROUP BY podcast_id) g ON p.podcast_id = g.podcast_id
GROUP BY author_id
HAVING COUNT(author_id) >= 100
ORDER BY 2 DESC
 ''', con)


author_id  Number_of_reviews  \
0   791F376ACA3C08D                613   
1   D3307ADEFFA285C                609   
2   96A3F0358E7808D                454   
3   04C3B6DEFE22FB1                431   
4   8858ED4249110B6                397   
5   7F9EA8B1A449BFC                353   
6   4B25015E4D7A1DA                340   
7   B20DB44F45DB97A                321   
8   A52EBC918BC19D1                311   
9   184F09732A68A08                308   
10  C02C514DAA9D885                306   
11  6033FF110F6CD23                299   
12  8EA31F030185A88                293   
13  3D53C30E1911630                287   
14  8843FBA0937364E                253   
15  8167974065A0260                244   
16  3E2C917B8E60AB3                192   
17  04E85BADC6E8986                184   
18  D42837C5F8BD234                177   
19  C0EA806D516803B                173   
20  675A82764372546                168   
21  61ED6D289B34165                164   
22  E1E3D5BF8BDF567                159   
23  B67D7BE0E454268                146   
24  871BCE7CB360EBB                145   
25  2975ECC1C888139                143   
26  01B41DDC59E7F7F                136   
27  59223EEA42FBFCE                133   
28  B3BCE599817B315                127   
29  CBF46E5BEDF2F9A                125   
30  6DBBFD322D38BD2                125   
31  426DDBF8D527E3A                124   
32  FB2467731CA599C                123   
33  E97FCC8DA48F89E                123   
34  C6ACA659087DCC1                123   
35  DCD872F1EE93AD5                121   
36  B53119FB3E276D3                119   
37  2899EB23FFF842C                116   
38  961F6822AAA4F4D                115   
39  304D4EE7DF74535                115   
40  B1877C1A2DEDAB4                110   
41  44C5207F0E59BA5                108   
42  241D91E70A625DB                108   
43  CBFA7DE2DA5D1D1                107   
44  C9EDC715E2AEA2F                107   
45  94F11F85A9AD087                106   
46  1AED5583C51E014                105   
47  E32005224E7097C                103   
48  AC072D3BF1CB749                103   
49  29A468BE322892B                103   
50  DE6E19493E35CE1                101   

                                                Title  Average_rating  
0                        Reality Life with Kate Casey            4.15  
1                     Transformational Leader Podcast            5.00  
2                     Transformational Leader Podcast            5.00  
3                                     Beyond Trending            5.00  
4                               KontroledKaos Podcast            5.00  
5                  Rich Hill Christian Church Podcast            5.00  
6                                 Capture The Culture            5.00  
7                               The Game Changer Life            5.00  
8         Christopher Lochhead Follow Your Different™            5.00  
9                                   Nine-Five Podcast            5.00  
10                              The Game Changer Life            5.00  
11                              The Game Changer Life            5.00  
12                     From Chaos to Peace with Conny            5.00  
13              Stephanie Miller's Happy Hour Podcast            5.00  
14                         I'll Have A Double Podcast            5.00  
15                The Mina Kimes Show featuring Lenny            4.97  
16                          The Haunted Horrorstorian            5.00  
17               Histories, Mysteries, & Conspiracies            5.00  
18                                    Beyond Trending            5.00  
19                              Stranger by the Dozen            4.62  
20                                           TrueAnon            2.61  
21                                           Newscast            3.80  
22                          Leveling Up with Eric Siu            5.00  
23                                Video Marketing 2.0            5.00  
24                       

The authors may have left multiple reviews for various episodes of a podcast. However, the COUNT() function used in the query aggregates the number of reviews per author over the entire podcast title, not individual episodes. This is a limitation of the data set and the query

Reviews on specific time

In [47]:
df_reviews.head(2)

podcast_id  \
0  c61aa81c9b929a66f0c1db6cbe5d8548   
1  c61aa81c9b929a66f0c1db6cbe5d8548   

                                              title  \
0                               really interesting!   
1  Must listen for anyone interested in the arts!!!   

                                             content  rating        author_id  \
0  Thanks for providing these insights.  Really e...       5  F7E5A318989779D   
1  Super excited to see this podcast grow. So man...       5  F6BF5472689BD12   

                  created_at  
0  2018-04-24T12:05:16-07:00  
1  2018-05-09T18:14:32-07:00

In [48]:
df=pd.read_sql_query(
        '''
SELECT 
  (CASE 
    WHEN CAST(SUBSTR(created_at, 12, 3) AS NUMERIC) BETWEEN 0 AND 6 THEN 'NIGHT'
    WHEN CAST(SUBSTR(created_at, 12, 3) AS NUMERIC) BETWEEN 6 AND 8 THEN 'EARLY MORNING'
    WHEN CAST(SUBSTR(created_at, 12, 3) AS NUMERIC) BETWEEN 8 AND 12 THEN 'MORNING'
    WHEN CAST(SUBSTR(created_at, 12, 3) AS NUMERIC) BETWEEN 12 AND 18 THEN 'AFTERNOON'
    WHEN CAST(SUBSTR(created_at, 12, 3) AS NUMERIC) BETWEEN 18 AND 20 THEN 'EARLY EVENING'
    ELSE 'EVENING'
  END) AS Part_of_the_day,
  COUNT(rating) AS Number_of_reviews,
  AVG(rating) AS Average_score
FROM reviews
GROUP BY Part_of_the_day
ORDER BY 2 DESC
 ''', con)
df

Part_of_the_day  Number_of_reviews  Average_score
0       AFTERNOON             639965       4.656694
1         MORNING             496738       4.643752
2           NIGHT             307384       4.558032
3   EARLY MORNING             240708       4.631849
4   EARLY EVENING             188901       4.655116
5         EVENING             151637       4.584560

In [49]:
fig = px.bar(df, x='Part_of_the_day', y='Number_of_reviews', color='Average_score',
             color_continuous_scale='Viridis')
fig.show()

Findings:

We can see here the reviewers are mostly active in afternoons to give the ratings

# **4.Inferential Statastics**

**4.1 CI for podcasts analysis based on business-careers**




In [50]:
business_careers = pd.read_sql_query(
    '''
    WITH business_careers AS (
      SELECT podcast_id,
             category
      FROM categories
      WHERE category = 'business-careers'
    )
    SELECT DISTINCT r.content AS review,
           b.podcast_id AS ID,
           r.rating AS rating
    FROM business_careers b
    INNER JOIN reviews r
    ON r.podcast_id = b.podcast_id
    ''', con)
business_careers


review  \
0      Nil talks with so many great people and is a g...   
1         Love the podcast and love the book! Thank you!   
2      Very insightful and eye opening podcast with i...   
3      Brilliant interview with two visionaries offer...   
4      I look forward to this podcast every morning o...   
...                                                  ...   
53258  It always joyful and inspiring to listen to ea...   
53259  Ya Brad!  I almost gave up on this thing.  You...   
53260  I recently started listening to the “Find your...   
53261  Had a ton of going this this interview and lov...   
53262  Sal Daher is incredible!  He has a deep unders...   

                                     ID  rating  
0      da7e26f4fbac05a7cfe042d3b5094e8d       5  
1      da7e26f4fbac05a7cfe042d3b5094e8d       5  
2      da7e26f4fbac05a7cfe042d3b5094e8d       5  
3      da7e26f4fbac05a7cfe042d3b5094e8d       5  
4      da7e26f4fbac05a7cfe042d3b5094e8d       5  
...                                 ...     ...  
53258  d2a71ab81c5dfab6f29faa244a65cf66       5  
53259  cc451ec20dfe8f985c246cc71749917d       5  
53260  a36b7e62a9cfa04d72714e6e9c8de42b       5  
53261  a4f82b38b130653057d9fe6dbbdcf0e6       5  
53262  a4f82b38b130653057d9fe6dbbdcf0e6       5  

[53263 rows x 3 columns]

In [51]:
business_careers.describe()['rating']

count    53263.000000
mean         4.890994
std          0.562712
min          1.000000
25%          5.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: rating, dtype: float64

In [52]:
import scipy.stats as stats

mean = business_careers['rating'].mean()
stddev = business_careers['rating'].std()
n = len(business_careers)
t_critical = stats.t.ppf(0.975, n - 1) # two-tailed t-distribution
margin_of_error = t_critical * stddev / (n ** 0.5)

confidence_interval = (mean - margin_of_error, mean + margin_of_error)

confidence_interval

(4.886214814382617, 4.895772681627747)

"The 95% confidence interval for the mean rating of podcasts in the 'business-careers' category is (4.8862, 4.8958). This means that we are 95% confident that the true mean rating for podcasts in this category falls between 4.8862 and 4.8958. The average rating for the podcasts in this category is 4.8867, which falls within the confidence interval."

**4.2 Retreving art related podcats with different groups**

In [53]:
df_comb = pd.read_sql_query( '''
                            SELECT DISTINCT r.content AS review,
                            c.podcast_id AS ID,
                            c.category,
                            r.rating AS rating
                            FROM categories c
                            INNER JOIN reviews r
                            ON r.podcast_id = c.podcast_id
                            WHERE c.category IN ('arts-performing-arts', 'arts-visual-arts')
                                ''', con
                            )  
df_comb


review  \
0      Thanks for providing these insights.  Really e...   
1      Super excited to see this podcast grow. So man...   
2      How funny this podcast was to me - I feel the ...   
3      Every episode of Snap Judgement feels like a m...   
4                             That’s all, simply amazing   
...                                                  ...   
46447  I enjoyed your Pace and Patience episode. I am...   
46448                     Just awful. Biased hot garbage   
46449  Please do a Part II of your episode on film mu...   
46450  Joshua is very talented and his excellent desc...   
46451  I happened upon this podcast and love it so mu...   

                                     ID              category  rating  
0      c61aa81c9b929a66f0c1db6cbe5d8548  arts-performing-arts       5  
1      c61aa81c9b929a66f0c1db6cbe5d8548  arts-performing-arts       5  
2      f2970432aad718a000ebac10e48ae6b0  arts-performing-arts       2  
3      f2970432aad718a000ebac10e48ae6b0  arts-performing-arts       5  
4      f2970432aad718a000ebac10e48ae6b0  arts-performing-arts       4  
...                                 ...                   ...     ...  
46447  b4c5dd7846b50106420bb11b848d98c4      arts-visual-arts       5  
46448  c99253c57c787198792c85eb32b53cdb  arts-performing-arts       1  
46449  efdb3fd297027f90a26c87a2be14ab00  arts-performing-arts       5  
46450  efdb3fd297027f90a26c87a2be14ab00  arts-performing-arts       5  
46451  efdb3fd297027f90a26c87a2be14ab00  arts-performing-arts       5  

[46452 rows x 4 columns]

In [54]:
df_comb.groupby('category').describe()

rating                                             
                        count      mean       std  min  25%  50%  75%  max
category                                                                  
arts-performing-arts  30056.0  4.705583  0.908338  1.0  5.0  5.0  5.0  5.0
arts-visual-arts      16396.0  4.822701  0.697645  1.0  5.0  5.0  5.0  5.0

For 'arts-performing-arts':

Mean = 4.705583

Standard deviation = 0.908338

For 'arts-visual-arts':

Mean = 4.822701

Standard deviation = 0.697645

 The formula for the t-interval with n-1 degrees of freedom, where n is the sample size, is:

 **mean ± t-critical * (standard deviation / √n)**

 Where t-critical is the critical value from the t-distribution table for the desired confidence level and degrees of freedom.



In [55]:
import numpy as np
from scipy import stats

grouped = df_comb.groupby('category')
mean_performing_arts = grouped.get_group('arts-performing-arts').rating.mean()
mean_visual_arts = grouped.get_group('arts-visual-arts').rating.mean()

standard_deviation_performing_arts = grouped.get_group('arts-performing-arts').rating.std()
standard_deviation_visual_arts = grouped.get_group('arts-visual-arts').rating.std()

n_performing_arts = grouped.get_group('arts-performing-arts').rating.count()
n_visual_arts = grouped.get_group('arts-visual-arts').rating.count()

t_critical = stats.t.ppf(0.975, n_performing_arts-1)
confidence_interval_performing_arts = (mean_performing_arts - t_critical * (standard_deviation_performing_arts / (n_performing_arts**0.5)), 
                                        mean_performing_arts + t_critical * (standard_deviation_performing_arts / (n_performing_arts**0.5)))

t_critical = stats.t.ppf(0.975, n_visual_arts-1)
confidence_interval_visual_arts = (mean_visual_arts - t_critical * (standard_deviation_visual_arts / (n_visual_arts**0.5)), 
                                    mean_visual_arts + t_critical * (standard_deviation_visual_arts / (n_visual_arts**0.5)))

print(confidence_interval_performing_arts)
print(confidence_interval_visual_arts)


(4.695313457927002, 4.71585236586858)
(4.8120212846112755, 4.833380032783212)


Findings:

Based on the confidence intervals for both arts-performing-arts and arts-visual-arts, we can say that with a 95% level of confidence, the true mean rating for arts-performing-arts lies between 4.6953 and 4.7159. And the true mean rating for arts-visual-arts lies between 4.8120 and 4.8334. These intervals provide a range of values that we expect the true mean to fall into.

It appears that the mean ratings for both arts categories are generally high, with a mean rating of 4.7056 for arts-performing-arts and 4.8227 for arts-visual-arts. These results suggest that listeners generally have a positive perception of podcasts in these categories.


**4.3 Hypothesis testing for business-careers podcasts**.




In [56]:
business_careers.describe()['rating']

count    53263.000000
mean         4.890994
std          0.562712
min          1.000000
25%          5.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: rating, dtype: float64

Based on the statistics provided, it looks like the mean rating is 4.89, with a standard deviation of 0.56. Most of the ratings seem to be 5, with a minimum of 1 and a maximum of 5.

To perform a hypothesis test, we need to specify the null and alternative hypotheses and choose a test statistic and significance level. Some possible hypothesis tests could be:

Test the mean rating against a specified value:

H0: The mean rating is equal to a specified value (e.g. 4.5)

Ha: The mean rating is not equal to the specified value

Test the proportion of 5-star ratings against a specified value:

H0: The proportion of 5-star ratings is equal to a specified value (e.g. 0.5)

Ha: The proportion of 5-star ratings is not equal to the specified value

Once the hypotheses have been defined, a test statistic (e.g. z-score or p-value) can be calculated and compared against a significance level (e.g. alpha = 0.05) to make a conclusion about the hypotheses.

In [57]:
# Load the data into a numpy array
data = [4.890994, 0.562712, 53263, 5.000000]
mean, std, n, specified_value = data


z = (mean - specified_value) / (std / np.sqrt(n))

# Calculate the p-value
p_value = stats.norm.sf(np.abs(z)) * 2

# Compare the p-value to the significance level
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis")
else:
    print("Fail to reject the null hypothesis")


Reject the null hypothesis


Findings:

Based on the results of the hypothesis test, we reject the null hypothesis that the mean rating is equal to the specified value (e.g. 4.5). This means that there is sufficient evidence to suggest that the mean rating is significantly different from the specified value.

In other words, the results of the test indicate that the mean rating for the business careers is not equal to the specified value, and therefore we can conclude that the business careers have a different mean rating. This conclusion is made based on a significance level of alpha = 0.05, meaning that there is a less than 5% chance of obtaining this result if the null hypothesis were true.

**4.4 Testing if difference in population mean is significant**

In [58]:
df_comb.groupby('category').describe()

rating                                             
                        count      mean       std  min  25%  50%  75%  max
category                                                                  
arts-performing-arts  30056.0  4.705583  0.908338  1.0  5.0  5.0  5.0  5.0
arts-visual-arts      16396.0  4.822701  0.697645  1.0  5.0  5.0  5.0  5.0

A hypothesis test is being conducted to determine if there is a significant difference between the average score of podcasts in the category "arts-performing-arts" and "arts-visual-arts". The null hypothesis is that there is no difference between the two means, while the alternative hypothesis is that there is a significant difference. The significance level is set at 5%.

Null: There is no difference between the average score of arts-performing-arts and arts-visuals-arts podcast.

Alternative: There is significant difference in mean average score.

Significance level - 5%.

In [59]:
significance_level=f'(μ1 - μ2) = {round((4.70 - 4.83), 2)}'
f'significance_level is {significance_level}'



'significance_level is (μ1 - μ2) = -0.13'

In [60]:
import scipy.stats as sm

# Define the data for each category
arts_performing = df_comb.loc[df_comb.category == 'arts-performing-arts', "rating"]
arts_visual = df_comb.loc[df_comb.category == 'arts-visual-arts', "rating"]

# Perform the t-test
result = sm.stats.ttest_ind(arts_performing, arts_visual, equal_var=False)
print(result)

# Check the p-value
if result[1] < 0.05:
    print("There is significant difference in mean average score.")
else:
    print("There is no significant difference in mean average score.")



Ttest_indResult(statistic=-15.49415863500572, pvalue=5.387959099297568e-54)
There is significant difference in mean average score.


Findings:


The results of the t-test indicate that the p-value (5.39e-54) is less than the significance level (0.05), suggesting strong evidence against the null hypothesis. This means that there is a significant difference between the average scores of podcasts in the two categories.

In summary, based on the hypothesis test and its results, we can conclude that there is a significant difference in the mean average score between podcasts in the "arts-performing-arts" and "arts-visual-arts" categories.

# **5.Conclusion**

I used pandas and SQL to analyze iTunes podcasts and found that there are 20 categories available. 

Business-culture is the most popular category. 

I also found the most highly rated and least highly rated podcasts, as well as the most active reviewers and their number of reviews.

Through statistical analysis, I concluded that to reach a wide audience as a future podcaster, it would be best to focus on less serious topics and have a large number of episodes available.